### PR TITLE
feat: log messages from middleware to sentry

### LINKED ISSUE #504: ApiErrors from middlewares aren't reported to sentry
Despite `SentryMiddleware` being the outtermost/last middleware, it does not receive errors triggered from other middlewares.

See [actix-web#1051](https://github.com/actix/actix-web/issues/1051#issuecomment-527268861) for some discussion of this behavior.

These errors should be reported to sentry but it may be tricky to do so for all potential cases: e.g. carefully distinguishing between an ApiError from a handler that SentryMiddleware will report vs one from the middleware (especially amid the Future chains).

The more important ones to report are from `DbTransactionMiddleware` (from the `db.commit/rollback()` calls). A potential "good enough" fix here may be just adding special case reporting for those two.

### PR DESCRIPTION
**Pull request by jrconlin**
## Description

Actix currently discards any errors generated in the middleware. You are
supposed to pass them in the body or some similar sub-optimal means.

Not a super fan of this approach, but basically, this patch attaches a
the error to the request or response extensions bucket. Since sentry is
also a middleware and gets run on every response, it sweeps the
request/response extension buckets and reports any errors it finds.

This has the downside that some context may be lost from the error (e.g.
I STRONGLY suspect the backtrace will be useless)

WIP:
Need to verify all middleware errors are being stored to extensions. 

Describe these changes.

## Testing

`debug!` statements have been added in the sentry middleware to note if an error has been detected and logged. You can trigger one such error (Bad Hawk Id) by starting a server and trying to fetch from `http://localhost:8000/test`

## Issue(s)

Closes #504

### CODE REVIEW

**Comment by fzzzy** [comment]
I mean, it sounds gross because actix trying to keep everything immutable and rust trying to keep everything owned makes it kind of necessary. Can anyone even think of another way to implement this? I guess do the network request for logging an error at every site that generates an error, but that seems gross.

I think this is fine.

**Comment by tublitzed** [comment]
@jrconlin regarding the testing instructions with the `debug!` statements - that isn't going to help us in prod though unless we have logging set to debug, correct?

**Comment by fzzzy** [comment]
@tublitzed No, the errors will still be logged in sentry on production -- we just won't get the detailed logging about the process of logging the error into sentry on production.

I don't think we could even handle that many log messages on production anyway, so if we want to see those debug messages we'll have to do it on stage or locally. We'll still get sentry errors on prod.

File: src/error.rs [hunk 1]
```diff
@@ -74,6 +75,15 @@ pub enum ApiErrorKind {
 
     #[fail(display = "{}", _0)]
     Validation(#[cause] ValidationError),
+
+    #[fail(display = "Precondition Error: {}", _0)]
+    Precondition(String),
```
[the remark this answers is not in the document]
Comment (jrconlin) [reply src/error.rs:80]: Not really sure we do, other than capture what exactly the APIError was. I didn't want to do additional parsing on the message content, so I figured I'd just add an additional enum. For example, `Precondition` checks for several types of errors, including Hawk and Db, but identifying the actual source of the error can be tricky. I figured that rolling those into `Precondition` may help identify where the failure is happening as well as the error being hit. I could probably `Vec` the errors like you suggest, but I'm not sure if that might complicate things with how `extension_mut.insert::<T>()` wants to deal with things. (Would I have to extract the data, update it, then re-insert the modified Vec.

[the remark this answers is not in the document]
Comment (pjenvey) [reply src/error.rs:80]: The Vec's a little annoying around extensions but seems doable (ignore my changes in sentry.rs):

https://github.com/mozilla-services/syncstorage-rs/compare/bug/504...bug/504-capture-vec

One issue with storing the ApiError directly though: the Vec requires ownership of it and they're not `Clone`. Instead we could generate the sentry Event in save_for_capture, storing it instead (`Vec<Event>`).

(Alternatively we could capture it to sentry immediately at the save point. But maybe we don't have all the tags we'd like by then?)

[the remark this answers is not in the document]
Comment (jrconlin) [reply src/error.rs:80]: Hmm. Yeah, we wouldn't have all the tags, so I'd hold off reporting it until the clean up like we are now. I'll look at storing Events though.

**Comment by pjenvey** [comment]
Ok: in [bug/504-pjenvey](https://github.com/mozilla-services/syncstorage-rs/compare/bug/504-pjenvey) I'm able to persist the original backtraces.

`event_from_fail` only needs a reference to a `Fail` so we can create the Event w/ backtrace on the fly without any borrowing needed inside the extensions `Vec<Event>`.

a couple other bits:

- it works on actix-web `Error` because they're common in middleware
- we don't need this stuff in get_hawk_id because the callers of it do it
- we missed capturing Errors from db commit/rollback (probably the most important to capture). We don't have extensions on hand at that point (sigh), so I just report it immediately with what's on hand

File: src/web/middleware/sentry.rs [hunk 2]
```diff
@@ -72,9 +103,41 @@ where
...
+                }
+            };
+            // add the uri.path (which can cause influx to puke)
+            tags.extra.insert("uri.path".to_owned(), uri);
             match sresp.response().error() {
-                None => {}
+                None => {
+                    // Middleware errors are eaten by current versions of Actix. Errors are now added
+                    // to the extensions. Need to check both for any errors and report them.
+                    if let Some(events) = sresp.request().extensions().get::<Vec<Event<'static>>>()
```
Comment (pjenvey) [inline src/web/middleware/sentry.rs:126]: we could `remove()` instead of `get()` here to avoid the clones, but Event looks probably cheap anyway

**Comment by jrconlin** [comment]
refactored to new, squashed branch since there were a lot of merges inbetween that made stuff messy.

**Pull request merged**